python/ray/_private/ray_logging.py

```diff
@@ -253,3 +261,153 @@ def emit(self, data):
 
 
 global_worker_stdstream_dispatcher = WorkerStandardStreamDispatcher()
+
+
+# Regex for canonicalizing log lines.
+NUMBERS = re.compile(r"(\d+|0x[0-9a-fA-F]+)")
+
+# How many seconds of messages to buffer for log deduplicatoin.
+AGGREGATION_WINDOW_S = 3.0
+
+
+def _canonicalise_log_line(line):
+    # Remove words containing numbers or hex, since those tend to differ between
+    # workers.
+    return " ".join(x for x in line.split() if not NUMBERS.search(x))
+
+
+@dataclass
+class DedupState:
+    # Timestamp of the earliest log message seen of this pattern.
+    timestamp: int
+
+    # The number of un-printed occurrances for this pattern.
+    count: int
+
+    # Latest instance of this log pattern.
+    line: int
+
+    # Latest metadata dict for this log pattern.
+    metadata: dict
+
+    # Set of (ip, pid) sources which have emitted this pattern.
+    sources: Set[Tuple[str, int]]
+
+
+class LogDeduplicator:
+    def __init__(self):
+        # Buffer of up to AGGREGATION_WINDOW_S recent log patterns.
+        # This buffer is cleared if the pattern isn't seen within the window.
+        self.recent: Dict[str, DedupState] = {}
+
+    def deduplicate(self, lines: List[str], metadata: dict) -> List[str]:
+        """Rewrite a batch of lines to reduce duplicate log messages.
+
+        Args:
+            lines: The batch of lines from a single source.
+            metadata: Dict describing the data source.
+
+        Returns:
+            Batch of deduplicated lines.
+        """
+        global _log_dedup_warned
+        if not RAY_DEDUP_LOGS:
+            return lines
+
+        source = (metadata.get("ip"), metadata.get("pid"))
+        now = time.time()
+        output = []
+
+        # Decide which lines to emit and which to buffer.
+        for line in lines:
+            if RAY_TQDM_MAGIC in line:
+                output.append(line)
+                continue
+            dedup_key = _canonicalise_log_line(line)
+            if dedup_key in self.recent:
+                sources = self.recent[dedup_key].sources
+                sources.add(source)
+                if len(sources) > 1:
+                    state = self.recent[dedup_key]
+                    self.recent[dedup_key] = DedupState(
+                        state.timestamp,
+                        state.count + 1,
+                        line,
+                        metadata,
+                        sources,
+                    )
+                else:
+                    # Don't dedup messages from the same source, just print.
+                    output.append(line)
+            else:
+                self.recent[dedup_key] = DedupState(now, 0, line, metadata, {source})
+                output.append(line)
+
+        # Flush patterns from the buffer that are older than the aggregation window.
+        while self.recent:
+            if now - next(iter(self.recent.values())).timestamp > AGGREGATION_WINDOW_S:
```

Comment:
Looks like the if here is False, `dedup_key = next(iter(self.recent))` can be buggy?

Hmm I found if it is else, we just break. Maybe we can refactor code this way to make this part clearer? 

```
dedup_state = next(iter(self.recent.values())
if now - dedup.state.timestamp < AGGREGATION_WINDOW_S:
    break

dedup_key = next(iter(self.recent))
...
```